sangria-codegen/src/main/scala/com.mediative.sangria.codegen/Importer.scala

```diff
@@ -55,42 +60,69 @@ case class Importer(schema: Schema[_, _], document: ast.Document) {
       touch: Boolean,
       name: String,
       tpe: Type,
-      selection: Option[Tree.Selection] = None): Tree.Field = {
+      selection: Option[Tree.Selection] = None,
+      union: Seq[Tree.UnionSelection] = Seq.empty): Tree.Field = {
     if (touch)
       touchType(tpe)
-    Tree.Field(name, tpe, selection)
+    Tree.Field(name, tpe, selection, union)
   }
 
-  def generateSelections(selections: Seq[ast.Selection]): Tree.Selection =
-    selections.map(generateSelection).foldLeft(Tree.Selection.empty)(_ + _)
+  def generateSelections(
+      selections: Seq[ast.Selection],
+      typeConditions: Set[String] = Set.empty): Tree.Selection =
+    selections.map(generateSelection(typeConditions)).foldLeft(Tree.Selection.empty)(_ + _)
+
+  def generateSelection(typeConditions: Set[String])(node: ast.AstNode): Tree.Selection = {
+    def filteredByTypeCondition(namedType: Option[ast.NamedType])(
+        f: => Tree.Selection): Tree.Selection =
+      if (typeConditions.nonEmpty && !namedType.map(_.name).filter(typeConditions).isDefined)
+        Tree.Selection.empty
+      else
+        f
 
-  def generateSelection(node: ast.AstNode): Tree.Selection = {
     typeInfo.enter(node)
     val result = node match {
       case field: ast.Field =>
         typeInfo.tpe match {
           case Some(tpe) =>
             tpe.namedType match {
+              case union: UnionType[_] =>
+                val types = union.types.toList.map { tpe =>
+                  // Prepend the union type name to include and descend into fragment spreads
+                  val conditions = (union.name +: tpe.name +: tpe.interfaces.map(_.name)).toSet
+                  val selection  = generateSelections(field.selections, conditions)
+                  Tree.UnionSelection(tpe, selection)
+                }
+
+                Tree.Selection(
+                  Vector(generateField(touch = false, field.outputName, tpe, union = types)))
```

Comment:
I agree this is inconsistent. The reason was to make it clear that the first parameter refers to whether or not the type should be "marked for generation". Replaced with explicit calls to `touchType` in 7b0cc9a.